### PR TITLE
Added Options for display and serialisation (faster Push); general refactoring

### LINKED ISSUE #49: Performance: Speckle Serialiser takes too long to complete
#### Description:

Speckle_Serialiser is responsible for almost all computation time while performing the Push. We're talking several seconds per <10 `Structure.Bar`s. I suspect it's due to a limitation of how speckle does its "serialisation" together with the deep nesting of our objects, which probably Speckle wasn't designed for. 
SpeckleSerialiser does in fact expose a `maximumNestLevel` parameter, which we cannot afford to use as we need to send eveything. (Edit: changing it does not seem to do anything).

Already highlighted the issue to Dimitri. I need to produce more detailed profiling; based on that potentially raise an issue in their Github. 

In the meantime, perhaps #58 could help.

### LINKED ISSUE #50: Improve consistency of Dispatcher methods

### LINKED ISSUE #51: Add PushConfig option to toggle "detailed" visualisation for BHoMObjects
#### Description:

E.g. Bars are always extruded and represented with Meshes now. This is heavy and might impact larger models. Add option to toggle.

Important to allow speed on larger models. 
Due to the bug #49 this issue can be considered as a critical workaround ATM.

### LINKED ISSUE #59: Implement converts for Brep and Surfaces; clarify geometry convert strategy
##### Context: get a visualisable representation of Surfaces and Nurbs in SpeckleViewer.

Unlike Meshes, Lines and Points, for which we already are the converts, other types (like Breps) are not so simple to convert.

It would be an easy job if we could reference SpeckleCoreGeometryRhino, because that project contains all the converts from Rhino classes to Speckle classes.
However, the only nuget package for SpeckleCoreGeometry seems only to include the SpeckleCoreGeometryClasses, which are the base geometry types (SpeckleLine, SpecklePoint etc).
We could solve this by
1. Rewriting the convert ourselves (I just attempted to do it and it's definitely not recommendable)
2. Hard-copying the .dll in our repo.
3. Waiting for the Nuget package to be produced: https://github.com/speckleworks/SpeckleCoreGeometry/issues/22.

If we clarify the strategy for this one, we might as well just pass through rhino for all our geometry types, then use the speckle converter for them.
This could be quite advantageous for us, because we could maintain less convert methods (the BHGeom to Rhino and vice-versa are needed anyway).

### PR DESCRIPTION
**Pull request by alelom**
### Issues addressed by this PR

Closes #50
Closes #51 
Closes https://github.com/BHoM/Speckle_Toolkit/issues/57
Closes #49 
Closes #59

### Test files

You need to build https://github.com/BHoM/Test_Toolkit to see all components in the Test file.

Test file located in the repo:
`GitHub\Speckle_Toolkit\Test Scripts\#53-TestScript\#53-TestScript.gh`

### Changelog

- The Push is now ~50x faster ([rewrote Speckle Serialisation to keep it to the essentials](https://github.com/BHoM/Speckle_Toolkit/pull/53/files#diff-d721640297208d04bd0edd24b04905ec)).
- Simplification of the Push process
- Simplification and more consistency of the Create and Convert methods
- Added `SpecklePushConfig.UseSpeckleSerialiser` to toggle the level of detail of exported BHoMObject in their SpeckleViewer representation.
- Added `SpecklePushConfig.UseSpeckleSerialiser` option which is by default false.

### CODE REVIEW

**Comment by alelom** [comment]
@al-fisher I can't seem to be able to add @izzylys to the reviewers, does that work for you? She seems  to be among the contributors already. Maybe she misses some permissions?

**Comment by al-fisher** [comment]
Sorry @izzylys! Thought you were already in the BHoM Org! Just sent you invite 😄

**Comment by alelom** [comment]
In order to find a solution to our performance issue (#49), from here I branched to [#58-BranchFrom#46-MultithreadingTests
](https://github.com/BHoM/Speckle_Toolkit/tree/%2358-BranchFrom%2346-MultithreadingTests) to test the impact of Parallel processing when serialising with SpeckleSerialiser, which (predictabily) ended up being a zero-gain modification.

**Comment by alelom** [comment]
To improve performance, I tried [a "partial" speckleSerialisation](https://github.com/BHoM/Speckle_Toolkit/issues/49#issuecomment-582900041), which didn't help.
Added an option to choose serialisation method as explained [here](https://github.com/BHoM/Speckle_Toolkit/issues/49#issuecomment-582975075).

File: Speckle_Engine/Compute/SpeckleSerialisation.cs [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+                }
+                return returnDict;
+            }
+
+            if (!myObject.GetType().AssemblyQualifiedName.Contains("System"))
+                return Serialise(myObject, recursionDepth + 1, path);
+
+            return null;
+        }
+    }
```
Comment (alelom) [inline Speckle_Engine/Compute/SpeckleSerialisation.cs]: Re-write of the Speckle `Serialise` method in its "essential" parts

**Comment by alelom** [comment]
@al-fisher , @PaulPoinet , @izzylys This is now ready for review.

Just try the test script as usual. There might be some component missing but that's not important, just plug something else.

@al-fisher the CI complains because it cannot find the reference to the SpeckleCoreGeometry dll that I had to manually copy in the repo (until Dimitri will produce a Nuget). What's the procedure for cases like this?

**Comment by PaulPoinet** [comment]
- Push works perfectly
- Pull has a problem deriving from the updated version of Speckle that I have on my machine and that is different from the one that @alelom has (Dec 2019). 

The issue with the Pull is to do with deserialisation. In SpeckleCore we get `GetObjectSubtypeBetter` failing and throwing the exception:
> Could not find subtype of 'SpeckleObject' with discriminator 'BH.oM.Structure.Elements.Bar
which comes from the  fact that `StreamGetObjectsAsync()` throws exception on line 2677 when attempting
```cs
result_ = JsonConvert.DeserializeObject<ResponseObject>( responseData_, _settings.Value );
```

@alelom will raise an issue and possibly collaborate with @didimitrie to find out what needs to be done.

**Pull request merged**